### PR TITLE
Add AWS SSO authentication and option aws_home_dir (s3fs-fuse#1931)

### PR DESCRIPTION
**Pull request by VivienMoretto87**
This pull request add a new authentication AWS SSO. I have although add the option aws_home_dir to configure the home directory (only available for the authentication AWS SSO)

### CODE REVIEW

**Comment by VivienMoretto87** [comment]
I have add a new library nlohmann-json. For Ubuntu/Debian it is the package is  nlohmann-json3-dev. Can you tell me how to add into the workflow?

**Comment by VivienMoretto87** [comment]
I have pushed all modifications for the ci in the github. All compilations and tests seems to pass

**Comment by ggtakec** [comment]
@VivienMoretto87 This PR seems to contain some fix.
We are using awscli for testing, but the chnaging to awscli version 2 should be separated from this PR.
I will consider it as well.

**Comment by VivienMoretto87** [comment]
@ggtakec Thank you for the reply. Awscli v1 is depretaced and many features are not implemented in awscli 1. 
Aws sso is implemented only in awscli v2 and so I have to use this new version to test it.

**Comment by ggtakec** [comment]
Prepared PR(#1933) to change AWS CLI version1 to version2.
Try to rebase this PR once it is merged.
I think that it can avoid the part where the error occurs in this PR.

Following is an additional question:
Is it possible to implement without using json parser (nlohmann-json)?
I'm wondering if I should increase the dependencies just to read the config file.
I would also like to hear @gaul opinion.

**Comment by ggtakec** [comment]
@sqlbot
As you said, I think this is just an additional logic to capture the STS credentials. And this has no way to renew that token.

@VivienMoretto87
Are you thinking of using s3fs from a so-called desktop PC?

I think there is also a way to not incorporate this PR logic directly into the s3fs itself.
After #1927 will be merged, we can delegate the tokens needed by s3 to an external DSO(library).
And I have already created a sample DSO for the authentication logic provided by the AWS SDK.(This includes authentication logic(including STS) provided by the AWS SDK)
It is up to the user to decide whether to use STS, but could you use this instead?

File: src/s3fs_cred.cpp [hunk 1]
```diff
@@ -905,12 +1058,12 @@ bool S3fsCred::InitialS3fsCredentials()
...
         return true;
     }
-
+    
     // 1 - keys specified on the command line
     if(IsSetAccessKeys(AutoLock::NONE)){
         return true;
     }
-
+
```
Comment (gaul) [inline src/s3fs_cred.cpp:1066]: Remove whitespace changes.  Repeated several times.

File: src/s3fs_help.cpp [hunk 2]
```diff
@@ -61,6 +61,11 @@ static const char help_string[] =
     "      - if it is not specified bucket name (and path) in command line,\n"
     "        must specify this option after -o option for bucket name.\n"
     "\n"
+    "   aws_home_path (default=\"\")\n"
+    "      - specify the aws path to use in the case of SSO authentication\n"
+    "        ${HOME}/.aws is used by default"
+    "      - this implementation is only avaialble for sso authentication\n"
+    "\n"
```
Comment (gaul) [inline src/s3fs_help.cpp:68]: Add documentation to `doc/man/s3fs.1`.

File: test/integration-test-common.sh [hunk 3]
```diff
@@ -34,6 +34,7 @@
 # S3_ENDPOINT="us-east-1"            Specify region
 # TMPDIR="/var/tmp"                  Set to use a temporary directory different
 #                                    from /var/tmp
+# Think to add the root certficate of AWS in the file /tmp/keystore.pem to execute aws_cli command
```
Comment (gaul) [inline test/integration-test-common.sh:37]: Should this have a `TODO:` prefix?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -152,7 +152,7 @@ jobs:
       - name: Build
         run: |
           ./autogen.sh
-          PKG_CONFIG_PATH=/usr/local/opt/curl/lib/pkgconfig:/usr/local/opt/openssl/lib/pkgconfig ./configure CXXFLAGS='-std=c++03 -DS3FS_PTHREAD_ERRORCHECK=1'
+          PKG_CONFIG_PATH=/usr/local/opt/curl/lib/pkgconfig:/usr/local/opt/openssl/lib/pkgconfig ./configure CXXFLAGS='-std=c++11 -DS3FS_PTHREAD_ERRORCHECK=1'
```
Comment (gaul) [inline .github/workflows/ci.yml:155]: Previously we supported Ubuntu 16.04 and RHEL 6 to a degree #1354 which only supported C++03.  We have talked about C++11 in the context of using the AWS SDK #1068.  I have no objection to this change but want to make sure that it is a conscious one motivated by the json library requirement.

File: src/s3fs_cred.cpp [hunk 5]
```diff
@@ -23,18 +23,23 @@
...
 #include "string_util.h"
 #include "metaheader.h"
+#include "s3fs_util.h"
 
+using json = nlohmann::json;
 //-------------------------------------------------------------------
 // Symbols
 //-------------------------------------------------------------------
 #define	DEFAULT_AWS_PROFILE_NAME	"default"
+#define M_ToConstUCharPtr(p)  reinterpret_cast<const unsigned char*>(p)
```
Comment (gaul) [inline src/s3fs_cred.cpp:42]: This macro is unused -- is it necessary?

File: src/s3fs_cred.cpp [hunk 6]
```diff
@@ -804,6 +827,135 @@ int S3fsCred::CheckS3fsCredentialAwsFormat(const kvmap_t& kvmap, std::string& ac
...
+            sso_region = value;
+        }else if(key == "sso_account_id"){
+            sso_account_id = value;
+        }else if(key == "sso_role_name"){
+            sso_role_name = value;
+        }
+    }
+
+     // No sso configure for the current profile 
+    if(sso_start_url != ""){
```
Comment (gaul) [inline src/s3fs_cred.cpp:897]: Change to `!sso_start_url.empty()`.

File: src/s3fs_cred.cpp [hunk 7]
```diff
@@ -804,6 +827,135 @@ int S3fsCred::CheckS3fsCredentialAwsFormat(const kvmap_t& kvmap, std::string& ac
...
+
+        // Digest the json via SHA1
+        unsigned char digest[SHA_DIGEST_LENGTH];
+        std::string json_sso_string = json_sso.dump();
+        S3FS_PRN_INFO("AWS SSO found : sha1 json_sso_string %s", json_sso_string.c_str());
+        int n = json_sso_string.length();
+        char json_sso_char[n]; 
+        strcpy(json_sso_char, json_sso_string.c_str()); 
+        // Digest the json
+        SHA1((unsigned char*)&json_sso_char, strlen(json_sso_char), (unsigned char*)&digest);
```
Comment (gaul) [inline src/s3fs_cred.cpp:913]: Is the `json_sso_char` local necessary?  Can you just call:

```c++
SHA1((unsigned char*)json_sso_string.c_str(), json_sso_string.length(), (unsigned char*)&digest);   
```

File: src/s3fs_cred.cpp [hunk 8]
```diff
@@ -804,6 +827,135 @@ int S3fsCred::CheckS3fsCredentialAwsFormat(const kvmap_t& kvmap, std::string& ac
...
+        // Digest the json via SHA1
+        unsigned char digest[SHA_DIGEST_LENGTH];
+        std::string json_sso_string = json_sso.dump();
+        S3FS_PRN_INFO("AWS SSO found : sha1 json_sso_string %s", json_sso_string.c_str());
+        int n = json_sso_string.length();
+        char json_sso_char[n]; 
+        strcpy(json_sso_char, json_sso_string.c_str()); 
+        // Digest the json
+        SHA1((unsigned char*)&json_sso_char, strlen(json_sso_char), (unsigned char*)&digest);   
+        char mdString[SHA_DIGEST_LENGTH*2+1];
```
Comment (gaul) [inline src/s3fs_cred.cpp:914]: Call `s3fs_hex_lower` instead.

File: src/s3fs_cred.cpp [hunk 9]
```diff
@@ -804,6 +827,135 @@ int S3fsCred::CheckS3fsCredentialAwsFormat(const kvmap_t& kvmap, std::string& ac
...
+ }
+
+//
+// Read Sso Aws Credential Json file
+// Use the library nlohmann-json3-dev to parse json
+//
+bool S3fsCred::ReadSsoAwsCredentialFile(const std::string &filename, AutoLock::Type type)
+{
+   
+    // open cerdentials cache pase file
```
Comment (gaul) [inline src/s3fs_cred.cpp:937]: Typo: credentials, parse

File: src/s3fs_cred.cpp [hunk 10]
```diff
@@ -804,6 +827,135 @@ int S3fsCred::CheckS3fsCredentialAwsFormat(const kvmap_t& kvmap, std::string& ac
...
+
+    std::string accesskey = jf["Credentials"]["AccessKeyId"];
+    std::string secret = jf["Credentials"]["SecretAccessKey"];
+    std::string session_token = jf["Credentials"]["SessionToken"];
+
+    if(session_token.empty()){
+      S3FS_PRN_EXIT("AWS session token was expected but wasn't provided in aws/cli/cache file");
+    }else{
+        if(!SetAccessKeyWithSessionToken(accesskey.c_str(), secret.c_str(), session_token.c_str(), type)){
+            S3FS_PRN_EXIT("session token is invalid. Think to refresh your sso creedentials.\n Execute aws sso login --profile profile and aws sts get-caller-identity --profile profile");
```
Comment (gaul) [inline src/s3fs_cred.cpp:952]: Typo: credentials

Rephrase as:

```
S3FS_PRN_EXIT("Session token is invalid. Refresh SSO credentials via:\n"
              "aws sso login --profile profile\n"
              "aws sts get-caller-identity --profile profile");
```

File: src/s3fs_cred.cpp [hunk 11]
```diff
@@ -951,6 +1103,24 @@ bool S3fsCred::InitialS3fsCredentials()
...
     }
 
+    // 2b - check ${HOME}/.aws/config 
+    // find sso_start_url, sso_region, sso_account_id,sso_role_name for the selected profile
+    std::string sso_cache_key;
+    if(CheckSsoCacheKey(sso_cache_key)){
+        // Use the cli cache. Check aws-cli github v2 code SSOCredentialFetcher in credentials.py
+        std::string aws_credentials_sso = std::string(getpwuid(getuid())->pw_dir) + "/.aws/cli/cache/"+ sso_cache_key +".json";
+        if(ItSetAwsHomePath()){
+        // Override the aws_home_dir path
```
Comment (gaul) [inline src/s3fs_cred.cpp:1113]: Fix indentation.

File: src/s3fs_cred.cpp [hunk 12]
```diff
@@ -951,6 +1103,24 @@ bool S3fsCred::InitialS3fsCredentials()
...
+        std::string aws_credentials_sso = std::string(getpwuid(getuid())->pw_dir) + "/.aws/cli/cache/"+ sso_cache_key +".json";
+        if(ItSetAwsHomePath()){
+        // Override the aws_home_dir path 
+            aws_credentials_sso =  aws_home_dir + "/cli/cache/"+ sso_cache_key +".json";
+        }
+        
+        if(ReadSsoAwsCredentialFile(aws_credentials_sso, AutoLock::NONE)){
+            return true;
+        }else{
+            S3FS_PRN_EXIT("Sso configuration is invalid. Think to refresh your sso creedentials.\n Execute aws sso login --profile profile and aws sts get-caller-identity --profile profile");
```
Comment (gaul) [inline src/s3fs_cred.cpp:1120]: Use a constant for this repeated string.

File: src/s3fs_help.cpp [hunk 13]
```diff
@@ -61,6 +61,11 @@ static const char help_string[] =
     "      - if it is not specified bucket name (and path) in command line,\n"
     "        must specify this option after -o option for bucket name.\n"
     "\n"
+    "   aws_home_path (default=\"\")\n"
+    "      - specify the aws path to use in the case of SSO authentication\n"
+    "        ${HOME}/.aws is used by default"
+    "      - this implementation is only avaialble for sso authentication\n"
```
Comment (gaul) [inline src/s3fs_help.cpp:67]: Typo: available.

**Review by gaul — CHANGES_REQUESTED** [review]

**Comment by gaul** [comment]
Please resolve merge conflicts as well.

**Comment by darioackermann** [comment]
We'd love to see this merged. Do you have time to review if we take care of the conflicts?

**Comment by ggtakec** [comment]
@darioackermann 
Could you try this #1927 merged code(credlib/credlib_ops option) and external DSO library(=https://github.com/ggtakec/s3fs-fuse-awscred-lib)?
This combination(options and libraries) should work for AWS authentication.(but you have to build)

**Comment by darioackermann** [comment]
> @darioackermann Could you try this #1927 merged code(credlib/credlib_ops option) and external DSO library(=https://github.com/ggtakec/s3fs-fuse-awscred-lib)? This combination(options and libraries) should work for AWS authentication.(but you have to build)

Yes that works, however I'd prefer if this behaviour was avilable as one installable packages, i.e. without manual building steps required. I have worked a bit on the branch and opened #2165 , happy to discuss.

**Comment by gaul** [comment]
Superseded by #2165.

**Pull request closed without merging**